cpp/src/arrow/json/parser.cc

```diff
@@ -49,21 +54,83 @@ using internal::checked_cast;
 using util::string_view;
 
 template <typename... T>
-Status ParseError(T&&... t) {
+static Status ParseError(T&&... t) {
   return Status::Invalid("JSON parse error: ", std::forward<T>(t)...);
 }
 
-Status KindChangeError(Kind::type from, Kind::type to) {
-  auto from_name = Tag(from)->value(0);
-  auto to_name = Tag(to)->value(0);
-  return ParseError("A column changed from ", from_name, " to ", to_name);
+static Status KindChangeError(Kind::type from, Kind::type to) {
+  return ParseError("A column changed from ", Kind::Name(from), " to ", Kind::Name(to));
+}
+
+const std::string& Kind::Name(Kind::type kind) {
+  static const std::string names[] = {"null",   "boolean", "number",
+                                      "string", "array",   "object"};
+
+  return names[kind];
+}
+
+const std::shared_ptr<const KeyValueMetadata>& Kind::Tag(Kind::type kind) {
+  static const std::shared_ptr<const KeyValueMetadata> tags[] = {
+      key_value_metadata({{"json_kind", Kind::Name(Kind::kNull)}}),
+      key_value_metadata({{"json_kind", Kind::Name(Kind::kBoolean)}}),
+      key_value_metadata({{"json_kind", Kind::Name(Kind::kNumber)}}),
+      key_value_metadata({{"json_kind", Kind::Name(Kind::kString)}}),
+      key_value_metadata({{"json_kind", Kind::Name(Kind::kArray)}}),
+      key_value_metadata({{"json_kind", Kind::Name(Kind::kObject)}}),
+  };
+  return tags[kind];
+}
+
+static internal::Trie MakeFromTagTrie() {
+  internal::TrieBuilder builder;
+  for (auto kind : {Kind::kNull, Kind::kBoolean, Kind::kNumber, Kind::kString,
+                    Kind::kArray, Kind::kObject}) {
+    DCHECK_OK(builder.Append(Kind::Name(kind)));
+  }
+  auto name_to_kind = builder.Finish();
+  DCHECK_OK(name_to_kind.Validate());
+  return name_to_kind;
+}
+
+Kind::type Kind::FromTag(const std::shared_ptr<const KeyValueMetadata>& tag) {
+  static internal::Trie name_to_kind = MakeFromTagTrie();
```

Comment:
The trie can be a speed optimization in some cases, but for that you shouldn't build it every time you do a lookup. I don't know if this function is performance-critical (presumably not?) so I'll let you choose your strategy here.

Reply:
It's static; it will only build the trie the first time this function is called